### DISCUSSION TITLE
Automatic ETL creation?

### DISCUSSION THREAD

**Discussion by ItsRobbAllen** [ravendb/ravendb#14297] [Q&A] [answered]
It's a breeze to automatically add indexes. 

Is there an easy way to scan an assembly for `AddEtlOperation`s & `RavenConnectionString`s and have those automatically added as well?

My microservices rely on the ETLs to pump data back & forth and since we do them manually, they get missed from time to time (ok... every time).

**Comment by ayende** [comment]
No, we don't have anything there. But given that ETL is typically just simple objects (with the scripts as string), it is should be possible for you to just put them all in a single location / scan them, no?

Something like `AbstractEtlFactor` and you'll get all of those.

Note that the reason we do the scanning of assemblies is that we need to translate Linq expressions to index definitions, which is not something that I would like to push to the users. Something like producing the ETL operations is far simpler.

**Reply by ItsRobbAllen** [reply]
Thanks.

It makes sense, I just wanted to make sure I wasn't missing something.

I'm sure I'm misreading the documentation, but when I add a connection string that already exists, it just overwrites it, but if I try adding an ETL that already exists, it throws an exception. How do I check to see if the ETL exists first?

**Reply by ayende** [reply]
You can call `GetEtlOperation`, no?

**Reply by ItsRobbAllen** [reply]
Sorry, I never replied to this.

That was the call I was missing, that should work fine. Thanks.

**Comment by ItsRobbAllen** [comment, marked as answer]
So, I finally got back to getting the ETL creation automated (hence the long delay), and I cannot find in the code `GetEtlOperation`. I see in the source code, under `Raven.Client.Documents.Operations.ETL`, code for Add & code for Update. But the Update requires the Task Id, which I do not know how to get.

Is there a back-door way of finding the Task ID of a particular ETL?

**Reply by ayende** [reply]
You can do it like this:

```
var record = await store.Maintenance.Server.SendAsync(new GetDatabaseRecordOperation(store.Database));
                    long taskId = record.RavenEtls.Find(x=>x.Name == "foo").TaskId;
```

**Reply by ItsRobbAllen** [reply]
That solved my problem. I can now create or update my ETLs without issue.

Thank you.
